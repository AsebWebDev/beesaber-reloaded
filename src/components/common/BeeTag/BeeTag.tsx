import { MDBBadge, MDBTooltip } from 'mdb-react-ui-kit';
import styled from 'styled-components';

import tokens from '@/tokens';

import type { Bee } from '@/../sharedTypes';

const BeeIcon = styled.i`
  padding-right: 0.5rem;
`;

const TrashIcon = styled.i.attrs({ as: 'a' })`
  padding-left: 0.5rem;

  &:hover {
    color: ${tokens.color.red.main};
    cursor: pointer;
  }
`;

type Props = {
  bee: Bee;
  handleDelete: (bee: Bee) => Promise<void>;
  handleSelect: (bee: Bee) => void;
  isSelected: boolean;
};

const BeeTag = ({
  bee,
  handleDelete,
  handleSelect,
  isSelected,
}: Props): JSX.Element => {
  const { playerName } = bee;

  const tooltipContent = <p>{bee.playerId}</p>;

  const handelDeleteClick = (e: React.MouseEvent<HTMLElement>) => {
    e.preventDefault();
    e.stopPropagation();
    void handleDelete(bee);
  };

  return (
    <span
      key={`${bee.playerId}${bee.playerName}`}
      onClick={() => handleSelect(bee)}
    >
      <MDBTooltip tag="span" title={tooltipContent}>
        <MDBBadge color={isSelected ? 'info' : 'warning'}>
          <BeeIcon className="fab fa-forumbee" aria-hidden="true" />
          {playerName}

          <TrashIcon
            className="fas fa-trash"
            onClick={handelDeleteClick}
            aria-label={`delete ${playerName}`}
          />
        </MDBBadge>
      </MDBTooltip>
    </span>
  );
};

export default BeeTag;
