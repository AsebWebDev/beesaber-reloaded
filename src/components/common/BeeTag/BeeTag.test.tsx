import { fireEvent, render, screen } from '@testing-library/react';

import exampleBee from '@/testing/testData/exampleBee';

import BeeTag from './BeeTag';

describe('component/common/BeeTag', () => {
  it.each([false, true])('should match snapshot', (isSelected) => {
    const { container } = render(
      <BeeTag
        bee={exampleBee}
        isSelected={isSelected}
        handleDelete={jest.fn()}
        handleSelect={jest.fn()}
      />
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should call handleSelect when bee is clicked', () => {
    const mockHandleSelect = jest.fn();

    render(
      <BeeTag
        bee={exampleBee}
        isSelected={false}
        handleDelete={jest.fn()}
        handleSelect={mockHandleSelect}
      />
    );

    const badge = screen.getByText(exampleBee.playerName);

    fireEvent.click(badge);

    expect(mockHandleSelect).toHaveBeenCalledTimes(1);
  });

  it('should call handleDelete when trash icon is clicked', () => {
    const mockHandleDelete = jest.fn();

    render(
      <BeeTag
        bee={exampleBee}
        isSelected={false}
        handleDelete={mockHandleDelete}
        handleSelect={jest.fn()}
      />
    );

    const trashIcon = screen.getByLabelText(`delete ${exampleBee.playerName}`);

    fireEvent.click(trashIcon);

    expect(mockHandleDelete).toHaveBeenCalledTimes(1);
  });
});
