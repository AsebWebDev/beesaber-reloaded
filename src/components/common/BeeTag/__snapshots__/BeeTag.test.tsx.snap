// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component/common/BeeTag should match snapshot 1`] = `
.c0 {
  padding-right: 0.5rem;
}

.c1 {
  padding-left: 0.5rem;
}

.c1:hover {
  color: #ff0025;
  cursor: pointer;
}

<span>
  <span>
    <span
      class="badge bg-warning"
    >
      <i
        aria-hidden="true"
        class="c0 fab fa-forumbee"
      />
      example Bee
      <a
        aria-label="delete example Bee"
        class="c1 fas fa-trash"
      />
    </span>
  </span>
</span>
`;

exports[`component/common/BeeTag should match snapshot 2`] = `
.c0 {
  padding-right: 0.5rem;
}

.c1 {
  padding-left: 0.5rem;
}

.c1:hover {
  color: #ff0025;
  cursor: pointer;
}

<span>
  <span>
    <span
      class="badge bg-info"
    >
      <i
        aria-hidden="true"
        class="c0 fab fa-forumbee"
      />
      example Bee
      <a
        aria-label="delete example Bee"
        class="c1 fas fa-trash"
      />
    </span>
  </span>
</span>
`;
